File: src/Roguelike.Core/Core/Map/MapManager.cs
using Roguelike.Core.Map;
using System.Collections.Generic;
using System.Linq;

namespace Roguelike.Core
{
    /// <summary>
    /// Manages the state of the game map for a single run, including the
    /// player's current position and available paths.
    /// </summary>
    public class MapManager
    {
        public MapGraph CurrentMap { get; private set; }
        public int CurrentNodeId { get; private set; } = -1;

        /// <summary>
        /// Generates a new map using the provided seed.
        /// </summary>
        public void GenerateNewMap(int seed, Dictionary<RoomType, float> roomWeights = null, 
                                   float monsterStarRatio = 0.5f, float eliteStarRatio = 0.5f)
        {
            Dictionary<RoomType, int> intWeights = null;
            if (roomWeights != null)
            {
                intWeights = new Dictionary<RoomType, int>();
                foreach (var kv in roomWeights)
                {
                    intWeights[kv.Key] = (int)Math.Round(kv.Value);
                }
            }
            
            var generator = new MapGenerator(seed, intWeights, monsterStarRatio, eliteStarRatio);
            CurrentMap = generator.Generate();
        }

        /// <summary>
        /// Retrieves the Room object for the player's current position.
        /// </summary>
        public Room GetCurrentRoom()
        {
            return CurrentNodeId == -1 ? null : CurrentMap.Rooms[CurrentNodeId];
        }

        /// <summary>
        /// Gets a list of rooms the player can move to from their current position.
        /// </summary>
        public List<Room> GetPossibleNextNodes()
        {
            if (CurrentMap == null) return new List<Room>();

            if (CurrentNodeId == -1)
            {
                return CurrentMap.RoomsOnFloor(0).ToList();
            }

            var currentRoom = GetCurrentRoom();
            if (currentRoom == null) return new List<Room>();

            return currentRoom.Outgoing.Select(id => CurrentMap.Rooms[id]).ToList();
        }

        /// <summary>
        /// Attempts to move the player to a new node.
        /// </summary>
        /// <param name="nodeId">The ID of the destination room.</param>
        /// <returns>True if the move was valid and successful, false otherwise.</returns>
        public bool MoveToNode(int nodeId)
        {
            var possibleMoves = GetPossibleNextNodes();
            if (possibleMoves.Any(r => r.Id == nodeId))
            {
                CurrentNodeId = nodeId;
                return true;
            }
            return false;
        }
    }
}

File: python_roguelike/core/map/map_manager.py
from typing import Dict, List, Optional
from .map_graph import MapGraph, Room
from .map_generator import MapGenerator
from ...data.enums import RoomType


class MapManager:
    def __init__(self):
        self.current_map: Optional[MapGraph] = None
        self.current_node_id: int = -1

    def generate_new_map(self, seed: int,
                         room_weights: Optional[Dict[RoomType, float]] = None,
                         monster_star_ratio: float = 0.5,
                         elite_star_ratio: float = 0.5):
        int_weights = None
        if room_weights is not None:
            int_weights = {rt: int(round(v)) for rt, v in room_weights.items()}

        generator = MapGenerator(seed, int_weights, monster_star_ratio, elite_star_ratio)
        self.current_map = generator.generate()

    def get_current_room(self) -> Optional[Room]:
        if self.current_node_id == -1:
            return None
        return self.current_map.rooms.get(self.current_node_id)

    def get_possible_next_nodes(self) -> List[Room]:
        if self.current_map is None:
            return []

        if self.current_node_id == -1:
            return self.current_map.rooms_on_floor(0)

        current = self.get_current_room()
        if current is None:
            return []

        return [self.current_map.rooms[nid] for nid in current.outgoing if nid in self.current_map.rooms]

    def move_to_node(self, node_id: int) -> bool:
        possible = self.get_possible_next_nodes()
        if any(r.id == node_id for r in possible):
            self.current_node_id = node_id
            return True
        return False
